ui/ts/components/table.ts

```diff
@@ -0,0 +1,108 @@
+/// <reference path="../typings/mithriljs/mithril.d.ts" />
+//
+module Components {
+  "use strict";
+
+  /**
+   * Table component is used to display a set of comparable objects in a table
+   * with each row representing the data of a single object from the set.
+   */
+  export module Table {
+    import MithrilElement = _mithril.MithrilVirtualElement;
+
+    /**
+     * TableColumn describes a single column of a data table.
+     */
+    export interface TableColumn<T> {
+      /**
+       * title is the string that will display in the header row for this
+       * column.
+       */
+      title: string;
+      /**
+       * data is a function which accepts the data for the row of the table, and
+       * returns either a string or a mithril element to display in the table in
+       * this column for the given row.
+       */
+      data: (row: T) => string|MithrilElement;
+    }
+
+    /**
+     * TableData is the argument for creating a table component.
+     */
+    export interface TableData<T> {
+      /**
+       * columns is a function which returns a array of TableColumn objects
+       * describing the columns of the table. Each TableColumn must accept the
+       * same row data type as the TableData itself. 
+       *
+       * The columns in the table will be displayed in the same order as the
+       * returned array.
+       */
+      columns: () => TableColumn<T>[];
+      /* rows is a function that returns an array of row data for display in the
+       * table. 
+       *
+       * Rows will be displayed in the same order as the returned array.
+       */
+      rows: () => T[];
+    }
+
+    class Controller<T> {
+      data: TableData<T>;
+
+      constructor(data: TableData<T>) {
+        this.data = data;
+      }
+
+      /**
+       * renderHeaders returns a mithril element which contains the header row
+       * for the table.
+       */
+      renderHeaders(): MithrilElement {
+        let renderedCols = [];
```

Comment:
Well, jsPerf is apparently down so google results are useless, but i'm fairly certain that for loops are still considerably faster than even the native .map() and .forEach() array helpers; I always write them this way by default. Some transpilers, like coffeescript, always compile loops to this form.

I assume you don't like the presence of iterators?